lib/github.ts

```diff
@@ -9,44 +9,73 @@ export const getLatestReleaseForBranch = async (
     branch: string;
   },
 ): Promise<GitHubRelease | null> => {
+  // the goal of this function is to find the latest release for a given branch. 
+  // the github api does not provide an easy way to do this. you must get releases and commits separately and then compare them.
+  //
+  // both of these endpoints have paging and we also want to minimize the number of requests we make to github for rate limiting. 
+  // 
+  // the strategy here is a mix between performance and simplicity for maintenance of the code. 
+  // First, get all releases for the repo. Even if we don't need to view them all, get them all otherwise we may need to implement some complex caching strategy. 
+  // Next, page through commits for a branch and compare to all releases. 
+  // We are assuming there are less releases than commits, which is what makes this performant. 
+
   let latestRelease: GitHubRelease | null = null;
 
+  let githubReleases: GitHubRelease[] = [];
+
+  await api.getTagsWithGitHubReleases({
+    owner,
+    repo,
+    processReleases: async(releases) => {
+      githubReleases = githubReleases.concat(releases);
+      return true; // continue paging
+    }
+  });
+  
   await api.getCommitsForBranch({
     owner,
     repo,
     branch,
     processCommits: async (commits) => {
-      // Given a page of commits, we want to find the latest commit that has a tag associated with it and also a github release for that tag.
-
-      for (const commit of commits) {
-        const releaseForTag = await api.getTagsWithGitHubReleases({
-          owner,
-          repo,
-          processReleases: (githubReleases) => {
-            for (const release of githubReleases) {
-              if (release.tag.commit.sha === commit.sha) {
-                latestRelease = release;
-              }
-            }
-
-            return latestRelease;
-          },
-        });
-
-        if (releaseForTag) { // found a release for the commit. This is the latest release for the branch.
-          latestRelease = releaseForTag;
+      for (const githubRelease of githubReleases) {
+        for (const commit of commits) {
+          if (githubRelease.tag.commit.sha === commit.sha && !latestRelease) {
+            latestRelease = githubRelease;
+          }
         }
       }
 
-      const getNextPageOfCommits = latestRelease == null; // continue paging if we haven't found the latest release yet.
+      const getNextPage = latestRelease === null;
 
-      return getNextPageOfCommits;
-    },
-  });
+      return getNextPage
+    }
+  })
 
   return latestRelease;
 };
 
+export const getAllCommitsSinceGivenCommit = async({ api, owner, repo, branch, lastTagSha }: { api: GitHubApi, owner: string; repo: string; branch: string, lastTagSha: string | undefined }): Promise<GitHubCommit[]> => {
```

Comment:
I previously deleted this function in a previous commit but had to bring it back.